### PR TITLE
Improve Tenant-Key Behavior #990

### PR DESCRIPTION
**Pull request by prb112**
- Introduced the TenantKeyFileUtil
	- Write a generated key to a file
	- Read a key from a file used during processing
- Update the fhir-persistence-schema README.md
- Update Db2AddTablePartition to cleanup code
- Update Main.java to check for Multitenant Feature instead of an OR

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-schema/README.md [hunk 1]
```diff
@@ -129,6 +129,8 @@ Don't forget to copy the tenant-key secret generated by --allocate-tenant. This
 --allocate-tenant default
 ```
 
+Use `--tenant-key-file tenant.key.file` to direct the action to read the tenant-key from file.  If the file exists the tenant key (up to 44 characters is read from the file.  If the file does not exist, the generated tenantKey is written out to the file.
```
Comment (lmsurpre) [inline fhir-persistence-schema/README.md:132]:
```suggestion
Use `--tenant-key-file tenant.key.file` to direct the action to read the tenant-key from file.  If the file exists the tenant key (up to 44 characters) is read from the file.  If the file does not exist, the generated tenantKey is written out to the file.
```

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java [hunk 2]
```diff
@@ -858,42 +886,103 @@ protected void addTenantKey() {
...
+            tenantKey = getRandomKey();
+        }
 
         // The salt is used when we hash the tenantKey. We're just using SHA-256 for
         // the hash here, not multiple rounds of a password hashing algorithm. It's
         // sufficient in our case because we are using a 32-byte random value as the
         // key, giving 256 bits of entropy.
         final String tenantSalt = getRandomKey();
 
+        Db2Adapter adapter = new Db2Adapter(connectionPool);
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:965]: if we ever add to the MULTITENANT_FEATURE_ENABLED list, we'll need to update this line...but I think its fine for now

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java [hunk 3]
```diff
@@ -858,42 +886,103 @@ protected void addTenantKey() {
...
+        checkIfTenantNameAndTenantKeyExists(adapter, tenantName, tenantKey);
+
         // Open a new transaction and associate it with our connection pool. Remember
         // that we don't support distributed transactions, so all connections within
         // this transaction must come from the same pool
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:970]: maybe move these below the if-else block so its closer to the try-with-resource?

File: fhir-persistence-schema/src/test/resources/test/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Blocks the Addition of the test files into git
+**/*.key.nowrite
```
Comment (lmsurpre) [inline fhir-persistence-schema/src/test/resources/test/.gitignore:2]: why does it end in `.nowrite`?
I think the pattern should match the name given in the examples in the README:  `--tenant-key-file tenant.key.file`

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:965]: Yes, that's right, we have a smattering of those throughout.

File: fhir-persistence-schema/src/test/resources/test/.gitignore (hunk 4 above)
Comment (prb112) [reply fhir-persistence-schema/src/test/resources/test/.gitignore:2]: right... I chose to ignore the files (.nowrite), as in one of the tests, I actually write the file. I wanted to not block the rest of the test files.  in src/test/resources/test/

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java (hunk 3 above)
Comment (prb112) [reply fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:970]: resolving conversation so I can edit in another suggestion

File: fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java [hunk 5]
```diff
@@ -858,42 +886,103 @@ protected void addTenantKey() {
...
-        Db2Adapter adapter = new Db2Adapter(connectionPool);
-        logger.info("Allocating new tenant: " + tenantName + " [key=" + tenantKey + "]");
+        if (tenantKeyFileName == null) {
+            logger.info("Allocating new tenant: " + tenantName + " [key=" + tenantKey + "]");
+        } else {
+            logger.info("Allocating new tenant: " + tenantName + " [tenantKeyFileName=" + tenantKeyFileName + "]");
+        }
+
```
Comment (prb112) [inline fhir-persistence-schema/src/main/java/com/ibm/fhir/schema/app/Main.java:976]:
```suggestion

        // Open a new transaction and associate it with our connection pool. Remember
        // that we don't support distributed transactions, so all connections within
        // this transaction must come from the same pool
```

**Comment by prb112** [comment]
made updates per code review

File: fhir-persistence-schema/src/test/resources/test/.gitignore (hunk 4 above)
Comment (lmsurpre) [reply fhir-persistence-schema/src/test/resources/test/.gitignore:2]: ah, ok.  i guess i was thinking this was a higher-level directory and we were trying to help prevent people from accidentally checking in their keys.  now i understand its just for the test.

**Pull request merged**